_posts/2018-01-14-smaller-docker-images.md

```diff
@@ -0,0 +1,405 @@
+```
+
+And now inspect the history:
+
+```bash
+$ docker history node-multi-stage2
+81cdd113ed21        7 seconds ago       /bin/sh -c #(nop)  CMD ["index.js"]             0B
+111777be4de0        7 seconds ago       /bin/sh -c #(nop)  EXPOSE 3000                  0B
+f753ecac370f        2 minutes ago       /bin/sh -c #(nop) COPY dir:854d4e5ce19a993ec…   1.59MB
+b87c2ad8344d        10 days ago         /bin/sh -c #(nop)  CMD ["node"]                 0B
+<missing>           10 days ago         /bin/sh -c set -ex   && for key in     6A010…   4.17MB
+<missing>           10 days ago         /bin/sh -c #(nop)  ENV YARN_VERSION=1.3.2       0B
+<missing>           10 days ago         /bin/sh -c ARCH= && dpkgArch="$(dpkg --print…   56.9MB
+<missing>           10 days ago         /bin/sh -c #(nop)  ENV NODE_VERSION=8.9.4       0B
+<missing>           4 weeks ago         /bin/sh -c set -ex   && for key in     94AE3…   129kB
+<missing>           4 weeks ago         /bin/sh -c groupadd --gid 1000 node   && use…   335kB
+<missing>           4 weeks ago         /bin/sh -c set -ex;  apt-get update;  apt-ge…   324MB
+<missing>           4 weeks ago         /bin/sh -c apt-get update && apt-get install…   123MB
+<missing>           4 weeks ago         /bin/sh -c set -ex;  if ! command -v gpg > /…   0B
+<missing>           4 weeks ago         /bin/sh -c apt-get update && apt-get install…   44.6MB
+<missing>           4 weeks ago         /bin/sh -c #(nop)  CMD ["bash"]                 0B
+<missing>           4 weeks ago         /bin/sh -c #(nop) ADD file:1dd78a123212328bd…   123MB
+```
+
+Hurrah! Has the file size changed at all?
+
+```bash
+$ docker images | grep node-multi-stage2
+node-multi-stage2               latest              abbe2e6023e9        About a minute ago   678MB
+node-multi-stage                latest              813b579d5aa5        11 minutes ago       681MB
+```
+
+Yes, the three layers image is slightly smaller.
+
+Not bad if you consider that this is just a simple _"Hello World!"_ application.
+
+But the image is still big!
+
+Is there anything you can do to make it even smaller?
+
+## 2. Remove all the unecessary cruft from the container with distroless
+
+The current image ships Node.js as well as `yarn`, `npm`, `bash` and a lot of other executables. It's also based on Ubuntu. So you have a full fledge operating system with all its little binaries and utilities.
+
+You don't need any of those when you run your container. The only dependency you need is Node.js.
+
+You could remove everything but Node.js. But how?
+
+Fortunately Google had the same idea and came up with [GoogleCloudPlatform/distroless](https://github.com/GoogleCloudPlatform/distroless).
+
+As the description for the repository points out:
+
+> "Distroless" images contain only your application and its runtime dependencies. They do not contain package managers, shells any other programs you would expect to find in a standard Linux distribution.
```

Comment:
yeah, sure but its worth adding a caveat to say that things compile with node-gyp (for js) or any other C library that depends on some system provided or other shared library it won't necessarily be there, and you'll break. ideally you'd statically compile your app and `FROM scratch` which may be available depending on the language choice

Reply:
Good point. I guess this is related to muslc vs glibc. If the base image is glibc friendly there's no problem. If the base image is Alpine, the entire thing is not guaranteed to work since we move from muslc to glibc.

Reply:
I'm thinking... perhaps this actually deserves a section on its own. If you statically compile your app... there's no need for a base image at all.

Reply:
that all said, it's worth raising as a caveat with distroless that it lags behind the language versions, e.g. currently its got `Node.js 8.9.1` in it but `Node.js 8.9.4` is the current LTS and `Node.js 9.4.0` is current. in between 8.9.1 and 8.9.4 there have been several feature fixes and more importantly **security updates**.

Reply:
The other gotcha is that it's not versioned as far as I can tell... The image has no tag?

Reply:
there are tags, so you can pin, but not named ones

Reply:
gcr.io/distroless/nodejs

Reply:
Exactly, no named tags. My expectation as a dev is that I can do `gcr.io/distroless/nodejs:8.9.1`. But I can't.

Reply:
since this post is about making a smaller image. the official `node:8-alpine` is 23MB the `gcr.io/distroless/nodejs:latest` is 25.1 MB

Reply:
yes, with caveats, though. And I listed those at the end.
I suggest we leave Distroless as is. The tag thing is annoying, but it's outside the scope of the  
 article.